### PR TITLE
Define an envelope type for newly added Ethereum tx.

### PR DESCRIPTION
**Pull request by sirano11**
## Proposed changes

This PR define an envelope type for newly added Ethereum tx.
  - To accept and distinguish Klaytn tx type and Ethereum tx type, it define an envelop type.
    - TxEthTypeEnvelope : 0x78
    - TxAccessList : 0x7801
    - TxDynamicFee : 0x7802
  - In order to have the same value as Ethereum's txHash, TxTypeEthEnvelop is not used when creating a txHash.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

#1128

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/types/tx_internal_data_access_list.go [hunk 1]
```diff
@@ -311,6 +311,22 @@ func (t *TxInternalDataAccessList) SerializeForSign() []interface{} {
 	}
 }
 
+func (t *TxInternalDataAccessList) TxHash() common.Hash {
+	return prefixedRlpHash(byte(t.Type()), []interface{}{
+		t.ChainID,
```
Comment (jimni1222) [inline blockchain/types/tx_internal_data_access_list.go:316]: Do we need to set chainId of the transaction in the `func (s londonSigner) Hash(tx *Transaction) common.Hash` when chainId has nil or empty value?
If not set, signature can be invalid because chainId will be nil in TxHash but signature made with s.chainid i think?

File: blockchain/types/tx_internal_data.go [hunk 2]
```diff
@@ -44,7 +44,7 @@ const (
 	// If types other than <base type> are not useful, they are declared with underscore(_).
 	// Each base type is self-descriptive.
 	// TODO-Klaytn-AccessList: Change TxTypeAccessList to another seperated value.
```
Comment (jimni1222) [inline blockchain/types/tx_internal_data.go:46]: This can be deleted?

File: blockchain/types/tx_internal_data.go [hunk 3]
```diff
@@ -250,6 +253,10 @@ func (t TxType) IsEthTypedTransaction() bool {
 	return t == TxTypeAccessList || t == TxTypeDynamicFee
```
Comment (aidan-kwon) [inline blockchain/types/tx_internal_data.go:253]: How about using a more generic way such as `return t & 0xff00 == TxTypeEthEnvelope << 8` ?

File: blockchain/types/tx_internal_data.go [hunk 4]
```diff
@@ -55,6 +55,9 @@ const (
 	TxTypeBatch, _, _
 	TxTypeChainDataAnchoring, TxTypeFeeDelegatedChainDataAnchoring, TxTypeFeeDelegatedChainDataAnchoringWithRatio
 	TxTypeLast, _, _
```
Comment (jimni1222) [inline blockchain/types/tx_internal_data.go:57]: Do we need to move `TxTypeLast`?

**Comment by aeharvlee** [comment]
Not a scope of this PR but, we need to fix the test cases.

Please take a look [this commit](https://github.com/klaytn/klaytn/pull/1172/commits/09fc3a2fba2ece8b184c4bbe9d2c2a13d7713fb0). I fixed test cases 🙂  and you can apply this changes in this PR too!

File: blockchain/types/tx_internal_rlp_encode_test.go [hunk 5]
```diff
@@ -1430,7 +1434,7 @@ func testTxRLPEncodeFeeDelegatedCancelWithRatio(t *testing.T) {
 		tx.TxSignatures,
 	})
 	assert.Equal(t, nil, err)
-
+	assert.Equal(t, hash(txHashRLP.Bytes()), NewTx(tx).Hash())
```
Comment (jimni1222) [inline blockchain/types/tx_internal_rlp_encode_test.go:1437]:
```suggestion
	assert.Equal(t, hash(txHashRLP.Bytes()), NewTx(tx).Hash())
	assert.Equal(t, hash(senderTxHashRLP.Bytes()), NewTx(tx).SenderTxHash())
```
Maybe adding logic to check sednerTxHash value in each case?

File: blockchain/types/tx_internal_data_access_list.go (hunk 1 above)
Comment (jimni1222) [reply blockchain/types/tx_internal_data_access_list.go:316]: https://github.com/klaytn/klaytn/pull/1171/commits/b97c0192c7534b5d65fc3b7b41a3f661a7f4c0a9

File: blockchain/types/tx_internal_rlp_encode_test.go [hunk 6]
```diff
@@ -266,6 +275,10 @@ func testTxRLPEncodeDynamicFee(t *testing.T) {
 	})
 	assert.Equal(t, nil, err)
 
+	assert.Equal(t, hash(txHashRLP.Bytes()), NewTx(tx).Hash())
+	senderTxHash, isFDTx := NewTx(tx).SenderTxHash()
```
Comment (jimni1222) [inline blockchain/types/tx_internal_rlp_encode_test.go:279]: Sorry for make you confuse;; 
I think you can use `SenderTxHashAll` to check senderTxHash works.

**Comment by aidan-kwon** [comment]
@sirano11 Could you resolve the conflict?

File: blockchain/types/tx_internal_data_ethereum_access_list.go [hunk 7]
```diff
@@ -367,7 +383,7 @@ func (t *TxInternalDataAccessList) String() string {
 	} else {
 		to = fmt.Sprintf("%x", t.GetRecipient().Bytes())
 	}
-	enc, _ := rlp.EncodeToBytes(t)
```
Comment (hqjang-pepper) [inline blockchain/types/tx_internal_data_ethereum_access_list.go:370]: Why did you change this argument of EncodeToBytes (from t to tx)?

**Review by aeharvlee — DISMISSED** [review]

File: blockchain/types/tx_internal_data_ethereum_access_list.go (hunk 7 above)
Comment (sirano11) [reply blockchain/types/tx_internal_data_ethereum_access_list.go:370]: If it encodes a `t`, it only encode fields in txInternalDataAccessList. 

However, Wehn encoding the transaction payload(a.k.a Hex in String()), it needs to additional field like `ethEnvelope || txType || rlp(txInternalDataAccessList)`.

So I implemented encoding / decoding logic in transaction.go(exactly `TxInternalDataSerializer`) and it encodes transaction variable `tx` to get transaction payload.

Comment (hqjang-pepper) [reply blockchain/types/tx_internal_data_ethereum_access_list.go:370]: Oh I got it, thanks for your kind comment

**Review by hqjang-pepper — APPROVED** [review]

**Review by jimni1222 — APPROVED** [review]
LGTM =)

**Review by aeharvlee — APPROVED** [review]

**Review by hqjang-pepper — APPROVED** [review]

File: blockchain/types/transaction_signing.go [hunk 8]
```diff
@@ -335,15 +337,15 @@ func (s londonSigner) SignatureValues(tx *Transaction, sig []byte) (R, S, V *big
 // Hash returns the hash to be signed by the sender.
 // It does not uniquely identify the transaction.
 func (s londonSigner) Hash(tx *Transaction) common.Hash {
-	if tx.Type() != TxTypeDynamicFee {
+	if tx.Type() != TxTypeEthereumDynamicFee {
 		return s.eip2930Signer.Hash(tx)
 	}
 
+	// infs[0] always has chainID
```
Comment (jeongkyun-oh) [inline blockchain/types/transaction_signing.go:344]: minor but this comment is not necessary here anymore.

**Review by jeongkyun-oh — APPROVED** [review]

**Review by aeharvlee — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]
